File: project-2-coldstart/js/src/service.mjs
// Service layer: validation + business rules over the store.
import * as store from "./store.mjs";

export function cleanTags(tags) {
  const cleaned = [];
  for (const raw of tags ?? []) {
    const tag = raw.trim().toLowerCase();
    if (!tag) throw new Error("tags must be non-empty strings");
    if (!cleaned.includes(tag)) cleaned.push(tag);
  }
  return cleaned;
}

export function createTask(title, tags = [], dbPath) {
  title = title.trim();
  if (!title) throw new Error("title must be non-empty");
  return store.addTask(title, cleanTags(tags), dbPath);
}

export function listTasks(tag = null, dbPath) {
  if (tag !== null) {
    tag = tag.trim().toLowerCase();
    if (!tag) throw new Error("tag filter must be non-empty");
  }
  return store.listTasks(tag, dbPath);
}

export function completeTask(id, dbPath) {
  return store.setDone(id, true, dbPath);
}

export function openTasks(tag = null, dbPath) {
  return listTasks(tag, dbPath).filter((t) => !t.done);
}

File: project-2-coldstart/python/tasks/service.py
"""Service layer: validation + business rules over the store."""
from tasks import store


def _clean_tags(tags):
    cleaned = []
    for tag in tags or []:
        tag = tag.strip().lower()
        if not tag:
            raise ValueError("tags must be non-empty strings")
        if tag not in cleaned:
            cleaned.append(tag)
    return cleaned


def create_task(title, tags=(), db_path=None):
    title = title.strip()
    if not title:
        raise ValueError("title must be non-empty")
    kw = {"db_path": db_path} if db_path else {}
    return store.add_task(title, _clean_tags(tags), **kw)


def list_tasks(tag=None, db_path=None):
    kw = {"db_path": db_path} if db_path else {}
    if tag is not None:
        tag = tag.strip().lower()
        if not tag:
            raise ValueError("tag filter must be non-empty")
    return store.list_tasks(tag=tag, **kw)


def complete_task(task_id, db_path=None):
    kw = {"db_path": db_path} if db_path else {}
    return store.set_done(task_id, True, **kw)


def open_tasks(tag=None, db_path=None):
    return [t for t in list_tasks(tag=tag, db_path=db_path) if not t["done"]]
